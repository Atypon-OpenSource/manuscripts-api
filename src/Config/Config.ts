/*!
 * Â© 2020 Atypon Systems LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as path from 'path'
import { load } from 'dotenv-safe'
import { isString, isNumber } from '../util'
import { CreateBucketOptions } from 'couchbase'
import {
  clientApplicationsFromSplitString,
  scopeConfigurationsFromSplitString,
} from '../Models/ClientApplicationModels'
import { existsSync, readFileSync } from 'fs'
import { log } from '../Utilities/Logger'
import { ConfigurationError } from '../Errors'
import {
  APIConfiguration,
  DatabaseConfiguration,
  AuthConfiguration,
  AWSConfiguration,
  GoogleConfiguration,
  EmailConfiguration,
  GatewayConfiguration,
  ServerConfiguration,
  ClientApplicationsConfiguration,
  EnvironmentLike,
  Environment,
  ConfigurationContainer,
  IAMConfiguration,
  ScopedAccessTokenConfiguration,
  LiteratumConfiguration,
  PressroomConfiguration,
  TemplateConfiguration,
  ShacklesConfiguration,
} from './ConfigurationTypes'
import { normalizeURL } from './normalize-url'

const isUTF8 = require('is-utf8')

/**
 * Configuration string validator:
 * passes input parameter through as return value
 * if it is a non-empty string, throws otherwise.
 */
function getString(value: any, key: string): string {
  if (!isString(value) || value.length === 0) {
    throw new ConfigurationError(key, value)
  }
  return value
}

function getArray(value: string, key: string, separator: string): string[] {
  if (value === '') {
    return []
  }
  const arrayParts = value.split(separator)
  for (const part of arrayParts) {
    if (!isString(part) || part.length === 0) {
      throw new ConfigurationError(key, value)
    }
  }
  return arrayParts
}

function getOptionalArray(envVar: string | undefined, key: string, separator: string): string[] {
  if (envVar === undefined || envVar === '') {
    return []
  }
  const value = getString(envVar, key)
  const arrayParts = value.split(separator)
  for (const part of arrayParts) {
    if (!isString(part) || part.length === 0) {
      throw new ConfigurationError(key, value)
    }
  }
  return arrayParts
}

// undefined intentionally here,
// in order to avoid creating an object with null values
// (which the relevant 3rd party code may not interpret similarly).
function getStringOptional(value: any): string | undefined {
  if (!isString(value) || value.length === 0) {
    return undefined
  }
  return value
}

/**
 * Configuration number validator:
 * passes input parameter through as return value
 * if it is a number, return null otherwise.
 *
 * The use of undefined (instead of null) is intentional, in order to avoid
 * passing objects with null values to 3rd party code (which may not interpret them correctly).
 */
function getNumberOptional(value: any): number | undefined {
  const numberVal = Number(value)
  if (!isNumber(numberVal) || isNaN(numberVal) || !isFinite(numberVal)) {
    return undefined
  }
  return numberVal
}

function getNumber(value: any, key: string, allowMissing?: boolean): number {
  const numberVal = Number(value)
  if (!allowMissing && (!isNumber(numberVal) || isNaN(numberVal) || !isFinite(numberVal))) {
    throw new ConfigurationError(key, value)
  }
  return numberVal
}

export class Configuration implements ConfigurationContainer {
  readonly API: APIConfiguration
  readonly DB: DatabaseConfiguration
  readonly auth: AuthConfiguration
  readonly AWS: AWSConfiguration
  readonly google: GoogleConfiguration
  readonly IAM: IAMConfiguration
  readonly email: EmailConfiguration
  readonly gateway: GatewayConfiguration
  readonly server: ServerConfiguration
  readonly apps: ClientApplicationsConfiguration
  readonly scopes: ScopedAccessTokenConfiguration[]
  readonly literatum: LiteratumConfiguration
  readonly pressroom: PressroomConfiguration
  readonly shackles: ShacklesConfiguration
  readonly template: TemplateConfiguration

  constructor(env: EnvironmentLike) {
    const bucketOptions = this.createBucketOptions(env)

    this.API = {
      port: Number(env.APP_PORT),
      oauthStateEncryptionKey: getString(
        env.APP_OAUTH_STATE_ENCRYPTION_KEY,
        'APP_OAUTH_STATE_ENCRYPTION_KEY'
      ),
      hostname: getString(env.APP_HOSTNAME_PUBLIC, 'APP_HOSTNAME_PUBLIC'),
    }

    /* istanbul ignore next */
    const skipVerification =
      env.NODE_ENV === 'development' &&
      Boolean(getNumber(env.APP_SKIP_ACCOUNT_VERIFICATION, 'APP_SKIP_ACCOUNT_VERIFICATION'))

    this.auth = {
      jwtSecret: getString(env.APP_JWT_SECRET, 'APP_JWT_SECRET'),
      skipVerification: skipVerification,
      hashSaltRounds:
        getNumber(env.APP_HASH_SALT_ROUNDS, 'APP_HASH_SALT_ROUNDS', true) ||
        (env.NODE_ENV === Environment.Production ? 10 : 3),
      serverSecret: getString(env.APP_SERVER_SECRET, 'APP_SERVER_SECRET'),
      enableNonConnectAuth: Boolean(
        getNumber(env.APP_ENABLE_NON_CONNECT_AUTH, 'APP_ENABLE_NON_CONNECT_AUTH')
      ),
    }

    const buckets = {
      user: getString(env.APP_USER_BUCKET, 'APP_USER_BUCKET'),
      data: getString(env.APP_DATA_BUCKET, 'APP_DATA_BUCKET'),
      project: getString(env.APP_DATA_BUCKET, 'APP_DATA_BUCKET'),
      state: getString(env.APP_STATE_BUCKET, 'APP_STATE_BUCKET'),
      derivedData: getString(env.APP_DERIVED_DATA_BUCKET, 'APP_DERIVED_DATA_BUCKET'),
    }

    // database initialization should not be happening without INITIALIZE_DATABASE='true',
    // except for when running tests when it should occur also when INITIALIZE_DATABASE was omitted.
    /* istanbul ignore next */
    const initialize =
      env.INITIALIZE_DATABASE === 'true' ||
      (typeof env.INITIALIZE_DATABASE === 'undefined' && env.NODE_ENV === Environment.Test)

    this.DB = {
      bucketOptions,
      buckets,
      initializeContents: initialize,
      uri: normalizeURL(getString(env.APP_DB_URI, 'APP_DB_URI')),
      username: getString(env.APP_COUCHBASE_ADMIN_USER, 'APP_COUCHBASE_ADMIN_USER'),
      password: getString(env.APP_COUCHBASE_ADMIN_PASS, 'APP_COUCHBASE_ADMIN_PASS'),
      bucketAdminPassword: getString(
        env.APP_COUCHBASE_RBAC_PASSWORD,
        'APP_COUCHBASE_RBAC_PASSWORD'
      ),
    }

    this.AWS = {
      accessKeyId: getString(env.APP_AWS_ACCESS_KEY_ID, 'APP_AWS_ACCESS_KEY_ID'),
      secretAccessKey: getString(env.APP_AWS_SECRET_ACCESS_KEY, 'APP_AWS_SECRET_ACCESS_KEY'),
      region: getString(env.APP_AWS_REGION, 'APP_AWS_REGION'),
    }

    this.google = {
      clientID: getString(env.APP_GOOGLE_CLIENT_ID, 'APP_GOOGLE_CLIENT_ID'),
      clientSecret: getString(env.APP_GOOGLE_CLIENT_SECRET, 'APP_GOOGLE_CLIENT_SECRET'),
      authCallback: normalizeURL(
        getString(env.APP_GOOGLE_AUTH_CALLBACK, 'APP_GOOGLE_AUTH_CALLBACK')
      ),
    }

    this.IAM = {
      clientID: getString(env.APP_IAM_CLIENT_ID, 'APP_IAM_CLIENT_ID'),
      authServerURL: getString(env.APP_IAM_SERVER_URL, 'APP_IAM_SERVER_URL'),
      authServerPermittedURLs: Array.from(
        new Set(
          getArray(
            getString(env.APP_IAM_PERMITTED_SERVER_URLS, 'APP_IAM_PERMITTED_SERVER_URLS'),
            'APP_IAM_PERMITTED_SERVER_URLS',
            ';'
          )
        )
      ),
      authCallbackPath: normalizeURL(
        getString(env.APP_IAM_AUTH_CALLBACK_PATH, 'APP_IAM_AUTH_CALLBACK_PATH')
      ),
      libraryURL: normalizeURL(getString(env.APP_IAM_LIBRARY_URL, 'APP_IAM_LIBRARY_URL')),
      apiServerURL: Array.from(
        new Set(
          getArray(
            getString(env.APP_API_SERVER_URL, 'APP_API_SERVER_URL'),
            'APP_API_SERVER_URL',
            ';'
          )
        )
      ),
    }

    this.email = {
      fromAddress: getString(env.APP_FROM_EMAIL, 'APP_FROM_EMAIL'),
      fromBaseURL: normalizeURL(getString(env.APP_BASE_URL, 'APP_BASE_URL')),
    }

    this.gateway = {
      cookieDomain: getString(env.APP_GATEWAY_COOKIE_DOMAIN, 'APP_GATEWAY_COOKIE_DOMAIN'),
      hostname: getString(env.APP_GATEWAY_HOSTNAME, 'APP_GATEWAY_HOSTNAME'),
      ports: {
        admin: getString(env.APP_GATEWAY_ADMIN_PORT, 'APP_GATEWAY_ADMIN_PORT'),
        public: getString(env.APP_GATEWAY_PUBLIC_PORT, 'APP_GATEWAY_PUBLIC_PORT'),
      },
    }

    const host = this.email.fromBaseURL.replace(/https{0,1}\:\/\//, '')
    const additionalOrigins = [`http://${host}`, `https://${host}`]
    this.server = {
      storeOnlySSLTransmittedCookies: Boolean(
        getNumber(
          env.APP_STORE_ONLY_SSL_TRANSMITTED_COOKIES,
          'APP_STORE_ONLY_SSL_TRANSMITTED_COOKIES'
        )
      ),
      allowedCORSOrigins: Array.from(
        new Set(
          getArray(
            getString(env.APP_ALLOWED_CORS_ORIGINS, 'APP_ALLOWED_CORS_ORIGINS'),
            'APP_ALLOWED_CORS_ORIGINS',
            ';'
          ).concat(additionalOrigins)
        )
      ), // get unique values from potentially duplicated ones.
    }

    this.literatum = {
      allowedIPAddresses: Array.from(
        new Set(
          getArray(
            getString(env.APP_ALLOWED_IP_ADDRESSES, 'APP_ALLOWED_IP_ADDRESSES'),
            'APP_ALLOWED_IP_ADDRESSES',
            ';'
          )
        )
      ),
    }

    this.apps = {
      knownClientApplications: clientApplicationsFromSplitString(
        getString(env.APP_CLIENT_APPLICATIONS, 'APP_CLIENT_APPLICATIONS'),
        ';',
        ','
      ),
    }

    this.scopes = scopeConfigurationsFromSplitString(
      getString(env.APP_CONTAINER_TOKEN_SCOPES, 'APP_CONTAINER_TOKEN_SCOPES'),
      ';',
      ','
    )

    this.pressroom = {
      baseurl: getString(env.APP_PRESSROOM_BASE_URL, 'APP_PRESSROOM_BASE_URL'),
      apiKey: getString(env.APP_PRESSROOM_APIKEY, 'APP_PRESSROOM_APIKEY'),
    }

    this.shackles = {
      baseUrl: getString(env.SHACKLES_URL, 'SHACKLES_URL'),
    }

    this.template = {
      allowedOwners: getOptionalArray(
        env.APP_COUCHBASE_ALLOWED_OWNERS,
        'APP_COUCHBASE_ALLOWED_OWNERS',
        ';'
      ),
      allowedProjects: getOptionalArray(
        env.APP_COUCHBASE_ALLOWED_PROJECTS,
        'APP_COUCHBASE_ALLOWED_PROJECTS',
        ';'
      ),
    }
  }

  static fromEnv(envExamplePath: string) {
    if (existsSync('.env') && isUTF8(readFileSync('.env'))) {
      const dotenvResult = load({
        allowEmptyValues: false,
        sample: envExamplePath,
      })

      // hack to ignore the case of missing .env
      if (dotenvResult && dotenvResult.error) {
        throw dotenvResult.error
      }
    } else {
      log.error('.env is missing or is not a UTF8 encoded text file.')
    }

    return process.env // load() call above loads .env file into process.env
  }

  public createBucketOptions(env: NodeJS.ProcessEnv): CreateBucketOptions {
    const authType = getStringOptional(env.APP_DB_AUTH_TYPE)
    const bucketType = getStringOptional(env.APP_DB_BUCKET_TYPE)
    const ramQuotaMB = getNumberOptional(env.APP_DB_RAM_QUOTA_MB)
    const replicaNumber = getNumberOptional(env.APP_DB_REPLICA_NUMBER)

    const bucketOptions: CreateBucketOptions = {}
    if (authType) {
      bucketOptions.authType = authType
    }
    if (bucketType) {
      bucketOptions.bucketType = bucketType
    }
    if (ramQuotaMB) {
      bucketOptions.ramQuotaMB = ramQuotaMB
    }
    if (replicaNumber) {
      bucketOptions.replicaNumber = replicaNumber
    }

    return bucketOptions
  }
}

const env = Configuration.fromEnv(path.join(__dirname, `../../.env`))
export const config = new Configuration(env)
