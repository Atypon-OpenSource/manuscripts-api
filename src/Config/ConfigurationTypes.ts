/*!
 * Â© 2020 Atypon Systems LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { RSA_JWK } from 'pem-jwk'

import { ClientApplication } from '../Models/ClientApplicationModels'

export enum Environment {
  Test = 'test',
  Development = 'development',
  Production = 'production',
}

/**
 * Represents an API configuration.
 */
export type APIConfiguration = {
  readonly port: number
  readonly hostname: string
}

/**
 * Represents a database configuration.
 */
export enum BucketKey {
  User = 'user',
  Project = 'project',
}

export type DatabaseConfiguration = {
  readonly buckets: { [name in BucketKey]: string }
}

export type ScopedAccessTokenConfiguration = {
  readonly name: string
  readonly secret: string
  readonly publicKeyPEM: string | null
  readonly publicKeyJWK: RSA_JWK | null
  readonly expiry: number
  readonly identifier: string // maps to the "kid" property when represented in JWK.
}

/**
 * Represents authentication configuration.
 */
export type AuthConfiguration = {
  readonly jwtSecret: string
  readonly skipVerification: boolean
  readonly hashSaltRounds: number
  readonly serverSecret: string
  readonly enableNonConnectAuth: boolean
}

/**
 * Represents email sending configuration.
 */
export type EmailConfiguration = {
  readonly fromAddress: string
  readonly fromBaseURL: string
}

export type ClientApplicationsConfiguration = {
  readonly knownClientApplications: Array<ClientApplication>
}

export type ServerConfiguration = {
  readonly allowedCORSOrigins: ReadonlyArray<string>
}

export interface EnvironmentLike {
  [key: string]: string | undefined
}

export interface PressroomConfiguration {
  readonly baseurl: string
  readonly apiKey: string
}

export interface ConfigurationContainer {
  readonly API: APIConfiguration
  readonly DB: DatabaseConfiguration
  readonly auth: AuthConfiguration
  readonly email: EmailConfiguration
  readonly server: ServerConfiguration
  readonly apps: ClientApplicationsConfiguration
  readonly scopes: ScopedAccessTokenConfiguration[]
  readonly pressroom: PressroomConfiguration
}
