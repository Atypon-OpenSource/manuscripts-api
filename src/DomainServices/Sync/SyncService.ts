/*!
 * Â© 2020 Atypon Systems LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as HttpStatus from 'http-status-codes'
import request from 'request-promise-native'

import { UserProfile, ObjectTypes } from '@manuscripts/manuscripts-json-schema'
import { v4 as uuid_v4 } from 'uuid'
import { randomBytes } from 'crypto'
import { promisify } from 'util'

import { SyncError, ValidationError, GatewayInaccessibleError } from '../../Errors'
import { BucketKey } from '../../Config/ConfigurationTypes'
import {
  appDataAdminGatewayURI, appDataAdminGatewayBaseURI,
  appDataPublicGatewayBaseURI
} from '../../Config/ConfigAccessors'
import { UserStatus, User } from '../../Models/UserModels'
import { IUserStatusRepository } from '../../DataAccess/Interfaces/IUserStatusRepository'
import { ISyncService } from './ISyncService'
import { timestamp } from '../../Utilities/JWT/LoginTokenPayload'
import { UserService } from '../User/UserService'

const randomBytesPromisified = promisify(randomBytes)

/**
 * Number of bytes for the randomly generated sync_gateway user account passwords
 */
export const SYNC_GATEWAY_PASSWORD_BYTE_COUNT = 40
/**
 * Name of the cookie sync_gateway is expecting
 */
export const SYNC_GATEWAY_COOKIE_NAME = 'SyncGatewaySession'
/**
 * Default expiry of the sync_gateway cookie, so we send it to the client with
 * the same expiry
 */
export const SYNC_GATEWAY_COOKIE_EXPIRY_IN_MS = 24 * 60 * 60 * 1000

export const GATEWAY_BUCKETS: Array<BucketKey> = [
  BucketKey.Data,
  BucketKey.DerivedData,
  BucketKey.Discussions
]

const randomPassword = () =>
  randomBytesPromisified(SYNC_GATEWAY_PASSWORD_BYTE_COUNT)
    .then(buf => buf.toString('hex'))

export const username = (userId: string) => {
  if (!userId.startsWith('User|')) {
    throw new ValidationError(`Invalid userId (${userId})`, userId)
  }
  return userId.replace('|', '_')
}

export class SyncService implements ISyncService {
  constructor (
    private userStatusRepository: IUserStatusRepository
  ) {}

  public static async isAlive () {
    return Promise.all([SyncService.isAdminInterfaceAlive(), SyncService.isPublicInterfaceAlive()])
  }

  private static async isAdminInterfaceAlive (): Promise<boolean> {
    const url = appDataAdminGatewayBaseURI()

    const options = {
      url: url,
      method: 'HEAD',
      resolveWithFullResponse: true
    }

    let response: any
    try {
      response = await request(options)
    } catch (error) {
      throw new GatewayInaccessibleError(url)
    }
    if (response.statusCode === HttpStatus.OK) {
      return true
    } else {
      throw new SyncError(`Header request to URL ${url} is failing`, response)
    }
  }

  private static async isPublicInterfaceAlive (): Promise<boolean> {
    const url = appDataPublicGatewayBaseURI()

    const options = {
      url: url,
      method: 'HEAD',
      resolveWithFullResponse: true
    }

    let response: any
    try {
      response = await request(options)
    } catch (error) {
      throw new GatewayInaccessibleError(url)
    }
    if (response.statusCode === HttpStatus.OK) {
      return true
    } else {
      throw new SyncError(`Header request to URL ${url} is failing`, response)
    }
  }

  public async removeGatewaySessions (
    userId: string,
    deviceId: string,
    userStatus: UserStatus
  ): Promise<boolean> {
    const { deviceSessions } = userStatus
    const sessionIDs = deviceSessions[deviceId]

    if (!sessionIDs || Object.values(sessionIDs).length === 0) {
      return false
    }

    const deletionResponses = await Promise.all(GATEWAY_BUCKETS.map(bucketKey => {
      const baseUrl = appDataAdminGatewayURI(bucketKey)
      const sessionId = sessionIDs[bucketKey]
      const options = {
        uri: `${baseUrl}/_user/${username(userId)}/_session/${sessionId}`,
        method: 'DELETE',
        json: true,
        simple: false,
        resolveWithFullResponse: true
      }
      try {
        const response = request(options)
        return response
      } catch (error) {
        throw new GatewayInaccessibleError(options.uri)
      }
    }))

    deletionResponses.forEach(response => {
      if ([HttpStatus.OK, HttpStatus.NOT_FOUND].indexOf(response.statusCode) < 0) {
        throw new SyncError(`Failed to delete SyncGateway session (${response.statusCode})`, response)
      }
    })

    delete deviceSessions[deviceId]

    await this.userStatusRepository.patchStatusWithUserId(userId, { deviceSessions }, {})

    // return `true` if all DELETE requests actually deleted something, `false` otherwise.
    return deletionResponses.every(response => response.statusCode === HttpStatus.OK)
  }

  public async gatewayAccountExists (userId: string, bucketKey: BucketKey): Promise<boolean> {
    const options = {
      method: 'GET',
      uri: `${appDataAdminGatewayURI(bucketKey)}/_user/${username(userId)}`,
      json: true,
      resolveWithFullResponse: true,
      simple: false
    }

    let response: any
    try {
      response = await request(options)
    } catch (error) {
      throw new GatewayInaccessibleError(options.uri)
    }
    if (response.statusCode === HttpStatus.OK) {
      return true
    } else if (response.statusCode === HttpStatus.NOT_FOUND) {
      return false
    } else {
      throw new SyncError(`Determining SyncGateway existence failed (${response.statusCode})`, response)
    }
  }

  // Creates a sync_gateway account
  public async createGatewayAccount (userId: string, bucketKey: BucketKey) {
    const sgUsername = username(userId)
    const options = {
      method: 'PUT',
      uri: `${appDataAdminGatewayURI(bucketKey)}/_user/${sgUsername}`,
      body: {
        password: await randomPassword(),
        admin_channels: [
          sgUsername,
          `${sgUsername}-readwrite`
        ],
        admin_roles: []
      },
      json: true,
      resolveWithFullResponse: true
    }

    let response: any
    try {
      response = await request(options)
    } catch (error) {
      throw new GatewayInaccessibleError(options.uri)
    }
    if (response.statusCode === HttpStatus.CREATED) {
      return sgUsername
    } else if (response.statusCode === HttpStatus.OK) {
      return sgUsername
    } else {
      throw new SyncError(`SyncGateway account upsert failed (${response.statusCode})`, response)
    }
  }

  // Creates a sync_gateway account
  public async createGatewayAdministrator (username: string, password: string, bucketKey: BucketKey, adminChannels: ReadonlyArray<string>, adminRoles: ReadonlyArray<string>) {
    const options = {
      method: 'PUT',
      uri: `${appDataAdminGatewayURI(bucketKey)}/_user/${username}`,
      body: {
        password: password,
        admin_channels: adminChannels,
        admin_roles: adminRoles
      },
      json: true,
      resolveWithFullResponse: true
    }
    let response: any
    try {
      response = await request(options)
    } catch (error) {
      throw new GatewayInaccessibleError(options.uri)
    }
    if (response.statusCode === HttpStatus.CREATED) {
      return
    } else if (response.statusCode === HttpStatus.OK) {
      return
    } else {
      throw new SyncError(`SyncGateway administrator account upsert failed (${response.statusCode})`, response)
    }
  }

  /**
   * removes a sync_gateway account.
   */
  public async removeGatewayAccount (userId: string) {
    await this.removeAllGatewaySessions(userId)

    const removeAccounts: request.RequestPromise[] = GATEWAY_BUCKETS.map((bucketKey) => {
      const uri = `${appDataAdminGatewayURI(bucketKey)}/_user/${username(userId)}`
      const options = {
        method: 'DELETE',
        uri: uri,
        json: true,
        resolveWithFullResponse: true,
        simple: false
      }
      let response: any
      try {
        response = request(options)
      } catch (error) {
        throw new GatewayInaccessibleError(options.uri)
      }
      return response
    })

    const responses = await Promise.all(removeAccounts)
    responses.forEach((response) => {
      if ([HttpStatus.OK, HttpStatus.NOT_FOUND].indexOf(response.statusCode) < 0) {
        throw new SyncError(`Failed to delete SyncGateway user (${response.url}: ${response.statusCode})`, response)
      }
    })
  }

  // Creates a session with a sync_gateway account
  // Removes existing sessions by default
  // TODO: rename this to make it clear it removes any existing session.
  public async createGatewaySessions (userId: string, deviceId: string, userStatus: UserStatus) {
    await this.removeGatewaySessions(userId, deviceId, userStatus)

    const sessionResponses = await Promise.all(
      GATEWAY_BUCKETS.map(async bucketKey => {
        const uri = `${appDataAdminGatewayURI(bucketKey)}/_session`
        const options = {
          method: 'POST',
          uri: uri,
          body: {
            name: username(userId)
          },
          json: true,
          simple: false,
          resolveWithFullResponse: true
        }

        let response: any
        try {
          response = await request(options)
        } catch (error) {
          throw new GatewayInaccessibleError(options.uri)
        }
        if (!response || !response.body || !response.body.session_id) {
          throw new SyncError(`Unable to parse session_id from response body (${JSON.stringify((response || {}).body)}`, response.body)
        }

        const sessionId = response.body.session_id as string

        return [bucketKey, sessionId] as [BucketKey, string]
      })
    )

    const { deviceSessions } = userStatus

    const sessionIds = sessionResponses
      .reduce((acc, [bucketKey, sessionId]) => {
        acc[bucketKey] = sessionId
        return acc
      }, {} as { [bucketKey in BucketKey]?: string })

    deviceSessions[deviceId] = sessionIds

    await this.userStatusRepository.patchStatusWithUserId(userId, { deviceSessions }, {})

    return deviceSessions[deviceId]
  }

  public async createGatewayContributor (user: User, bucketKey: BucketKey) {
    const uri = `${appDataAdminGatewayURI(bucketKey)}/`

    const [firstName] = user.name.split(' ', 1)
    const lastName = user.name.substring(firstName.length + 1)

    const userProfileId = UserService.profileID(user._id)

    const date = timestamp()

    const userProfile: UserProfile = {
      _id: userProfileId,
      objectType: ObjectTypes.UserProfile,
      userID: username(user._id),
      bibliographicName: {
        _id: `${ObjectTypes.BibliographicName}:${uuid_v4()}`,
        objectType: ObjectTypes.BibliographicName,
        given: firstName,
        family: lastName
      },
      email: user.email,
      createdAt: date,
      updatedAt: date
    }

    const options = {
      method: 'POST',
      uri,
      body: userProfile,
      json: true,
      resolveWithFullResponse: true,
      simple: false
    }

    let response: any
    try {
      response = await request(options)
    } catch (error) {
      throw new GatewayInaccessibleError(options.uri)
    }
    if (
      response.statusCode === HttpStatus.OK ||
      response.statusCode === HttpStatus.CREATED
    ) {
      return
    } else {
      throw new SyncError(
      `SyncGateway contributor creation failed`,
      response.body
    )
    }
  }

  public async removeAllGatewaySessions (userId: string): Promise<boolean> {
    const removeSessions = GATEWAY_BUCKETS.map(bucketKey => {
      const uri = `${appDataAdminGatewayURI(bucketKey)}/_user/${username(userId)}/_session`
      const options = {
        method: 'DELETE',
        uri: uri,
        json: true,
        resolveWithFullResponse: true,
        simple: false
      }
      let response: any
      try {
        response = request(options)
      } catch (error) {
        throw new GatewayInaccessibleError(options.uri)
      }
      return response
    })

    const responses = await Promise.all(removeSessions)
    responses.forEach(response => {
      if ([HttpStatus.OK, HttpStatus.NOT_FOUND].indexOf(response.statusCode) < 0) {
        throw new SyncError(`Failed to delete SyncGateway sessions (${response.url}: ${response.statusCode})`, response)
      }
    })

    await this.userStatusRepository.patchStatusWithUserId(userId, { deviceSessions: {} }, {})

    // return `true` if all DELETE requests actually deleted something, `false` otherwise.
    return responses.every(response => response.statusCode === HttpStatus.OK)
  }
}
