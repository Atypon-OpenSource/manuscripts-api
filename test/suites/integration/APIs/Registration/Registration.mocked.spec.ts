/*!
 * Â© 2020 Atypon Systems LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as HttpStatus from 'http-status-codes'
import * as supertest from 'supertest'

jest.mock('email-templates', () => jest.fn().mockImplementation(() => {
  return {
    send: jest.fn(() => Promise.resolve({})),
    render: jest.fn(() => Promise.resolve({}))
  }
}))

jest.mock('../../../../../src/DomainServices/External/AWS', () => ({
  SES: { sendEmail: jest.fn((_foo, callback) => callback(null, { foo: 1 })) }
}))

import { DIContainer } from '../../../../../src/DIContainer/DIContainer'
import { connectSignup, signup, verify } from '../../../../api'
import { TEST_TIMEOUT } from '../../../../utilities/testSetup'
import { drop, seed, testDatabase, dropBucket } from '../../../../utilities/db'
import { invalidNewUserCredentials, validNotVerifiedNewUserCredentials } from '../../../../data/fixtures/registrationCredentials'
import { ValidContentTypeAcceptJsonHeader, ValidHeaderWithApplicationKey } from '../../../../data/fixtures/headers'
import { SeedOptions } from '../../../../../src/DataAccess/Interfaces/SeedOptions'
import { BucketKey } from '../../../../../src/Config/ConfigurationTypes'
import { validBody } from '../../../../data/fixtures/credentialsRequestPayload'
import { GATEWAY_BUCKETS } from '../../../../../src/DomainServices/Sync/SyncService'

let db: any = null
const seedOptions: SeedOptions = { users: true, singleUseTokens: true, applications: true }

beforeAll(async () => db = await testDatabase())
afterAll(() => db.bucket.disconnect())

jest.setTimeout(TEST_TIMEOUT)

describe('UserRegistrationService - signup', () => {
  beforeEach(async () => {
    await drop()
    await dropBucket(BucketKey.Data)
    await seed(seedOptions)
  })
  test('should fail if user status does not exist', async () => {
    await DIContainer.sharedContainer.userStatusRepository.remove(null)
    const response: supertest.Response = await signup(
        invalidNewUserCredentials ,
        ValidContentTypeAcceptJsonHeader
      )

    expect(response.status).toBe(HttpStatus.UNAUTHORIZED)
  })

  test('should fail if verified user exists', async () => {
    const response: supertest.Response = await signup(
      invalidNewUserCredentials,
      ValidContentTypeAcceptJsonHeader
    )
    expect(response.status).toBe(HttpStatus.CONFLICT)
    expect(response.body).toMatchSnapshot()
  })

  test('should fail if not verified user exists', async () => {

    const response: supertest.Response = await signup(
      validNotVerifiedNewUserCredentials,
      ValidContentTypeAcceptJsonHeader
    )
    expect(response.status).toBe(HttpStatus.CONFLICT)
  })
})

describe('UserRegistrationService - verify', () => {
  beforeEach(async () => {
    await drop()
    await dropBucket(BucketKey.Data)
    await seed(seedOptions)
  })

  test('should fail if token is empty', async () => {
    const body = {
      token: ''
    }
    const response: supertest.Response = await verify(body)

    expect(response.status).toBe(HttpStatus.BAD_REQUEST)
  })

  test('should fail if token is not in db', async () => {
    const body = {
      token: 'not-in-db'
    }
    const response: supertest.Response = await verify(body)

    expect(response.status).toBe(HttpStatus.UNAUTHORIZED)
  })

  test('should verify user if token is valid', async () => {
    const body = {
      token: 'SingleUseToken|foobarbaz2'
    }
    const response: supertest.Response = await verify(body)

    expect(response.status).toBe(HttpStatus.NO_CONTENT)
  })
})

describe('UserRegistrationService - connectSignup', () => {
  beforeEach(async () => {
    await drop()
    await dropBucket(BucketKey.Data)
    await seed(seedOptions)
    return Promise.all(
        GATEWAY_BUCKETS.map(key => {
          return DIContainer.sharedContainer.syncService.createGatewayAccount(
              'User|' + validBody.email,
              key
          )
        })
    )
  })
  test('should fail if user email already exists', async () => {
    const response: supertest.Response = await connectSignup(
      {
        email: validBody.email,
        name: 'validName',
        connectUserID: 'validConnectId'
      }, {
        ...ValidHeaderWithApplicationKey
      }
    )
    expect(response.status).toBe(HttpStatus.CONFLICT)
  })
})
