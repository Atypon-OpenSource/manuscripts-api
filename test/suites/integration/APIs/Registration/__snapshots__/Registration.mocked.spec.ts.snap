// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserRegistrationService - signup should fail if verified user exists 1`] = `
Object {
  "error": "{\\"statusCode\\":409,\\"existingValue\\":{\\"_id\\":\\"User|valid-user@manuscriptsapp.com\\",\\"name\\":\\"Valid System User\\",\\"email\\":\\"valid-user@manuscriptsapp.com\\",\\"connectUserID\\":\\"valid-connect-user-id\\"},\\"name\\":\\"ConflictingRecordError\\"}",
}
`;
